File: Hashing/hashMapWithDoubleHashing.cpp
#include <bits/stdc++.h>
using namespace std;
using namespace std::chrono;

// Creating HashMap with quadratic Probing

// Double hashing is a collision resolution technique used in hash
//  tables. It works by using two hash functions to compute two
//  different hash values for a given key. The first hash function
//  is used to compute the initial hash value, and the second hash
//  function is used to compute the step size for the probing sequence.

/*
Double hashing can be done using :
(hash1(key) + i * hash2(key)) % TABLE_SIZE
Here hash1() and hash2() are hash functions and TABLE_SIZE
is size of hash table.
(We repeat by increasing i when collision occurs)
*/

/*

A popular second hash function is
hash2(key) = PRIME – (key % PRIME)
where PRIME is a prime smaller than the TABLE_SIZE.

*/

// hash Node Class

class HashNode
{
public:
    int value;
    int key;

    HashNode(int key, int value)
    {
        this->key = key;
        this->value = value;
    }
};

// HashMap Class

class HashMap
{

    // Hash elements array

    HashNode **arr;

    // hash capacity

    int capacity;

    // size
    int size;

    // counting collisions

    int collisions;

    // prime number less than the capacity

    int prime;

    // sieve of eratosthenes for marking all prime number less than the capcity

    vector<bool> isPrime;

public:
    // Constructor
    HashMap(int capacity)
    {
        this->capacity = capacity;
        size = 0;
        collisions = 0;
        // creating hashNode array
        arr = new HashNode *[capacity];

        // initializing all values with NULL

        for (int i = 0; i < capacity; i++)
        {
            arr[i] = NULL;
        }

        // seting prime value

        this->setSieve();

        prime = capacity - 1;
        while (isPrime[prime] != 1)
            prime--;
    }

    // print function

    void print(string message)
    {
        cout << endl
             << message;
    }

    // making sieve for prime numbers

    void setSieve()
    {

        vector<bool> temp(capacity, true);
        this->isPrime = temp;

        isPrime[0] = isPrime[1] = 1;
        for (long long i = 2; i * i <= this->capacity; i++)
            if (isPrime[i] == 1)
                for (long long j = i * i; j <= capacity; j += i)
                    isPrime[j] = 0;
    }

    // first hash function

    int hash1(int key)
    {
        return key % capacity;
    }

    // second hash function

    int hash2(int key)
    {
        return (prime - (key % prime));
    }

    // function for value insertion --

    void insertNode(int key, int value)
    {

        // if hashmap is full
        if (this->size >= capacity)
        {
            print("HashMap is Full");
            return;
        }

        HashNode *temp = new HashNode(key, value);

        int hashIndex = (hash1(key) + hash2(key)) % capacity;

        // finding the free space
        // bool if collision detected
        bool collisionExist = false;

        // Using Double hashing

        while (arr[hashIndex] != NULL && arr[hashIndex]->key != key)
        {
            collisionExist = true;
            // incrementing hashIndex and again making a hashcode for indexing
            hashIndex++;
            hashIndex = hashIndex % capacity;
        }

        if (collisionExist)
        {
            this->collisions++;
        }

        if (arr[hashIndex] == NULL)
        {
            // increasing the size of HashMap
            size++;
            arr[hashIndex] = temp;
        }

        // updating the value on particular key

        if (arr[hashIndex]->key == key)
        {
            arr[hashIndex]->value = value;
        }
    }

    // switch case function for insertion

    void insertForMenu()
    {
        // taking Input

        int key;
        print("Enter the Key: ");
        cin >> key;
        int value;
        print("Enter the value: ");
        cin >> value;

        insertNode(key, value);
    }

    // for deleting a key value pair

    int deleteNode(int key)
    {
        // (hash1(key) + i * hash2(key)) % TABLE_SIZE  --> HashIndex

        int hashIndex = (hash1(key) + hash2(key)) % capacity;

        // finding the node

        while (arr[hashIndex] != NULL)
        {
            if (arr[hashIndex] && arr[hashIndex]->key == key)
            {
                // removing the key value pair node
                HashNode *temp = arr[hashIndex];

                arr[hashIndex] = NULL;

                // decrementing size
                size--;

                return temp->value;
            }
            else
            {
                // if not found

                // using quadratic probing

                hashIndex++;
                hashIndex = hashIndex % capacity;
            }
        }
        cout << "endl"
             << "Key Not Found.";
        return 0;
    }

    // void function for deleting

    void deleteNodeVoid()
    {
        // Taking input from the user

        int key;
        print("Enter the Key: ");
        cin >> key;

        print("The key deleted with value : ");
        cout << deleteNode(key);
    }

    int getValueInMap(int key)
    {

        int hashIndex = (hash1(key) + hash2(key)) % capacity;
        // finding the key for value

        while (arr[hashIndex] != NULL)
        {
            if (arr[hashIndex] && arr[hashIndex]->key == key)
            {
                return arr[hashIndex]->value;
            }

            hashIndex++;
            hashIndex = hashIndex % capacity;
        }
        cout << endl
             << "Key Not found.";
        return 0;
    }

    // void get value

    void getValue()
    {
        // taking Input from the user

        int key;
        print("Enter the key to find the value: ");
        cin >> key;

        print("Value is : ");
        cout << getValueInMap(key);
    }

    // Getting a size

    int sizeOfMap()
    {
        return size;
    }

    // void function to get a size so that it does not return something menu can run soomthly

    void getSize()
    {
        print("The size of Map is : ");
        cout << sizeOfMap();
    }

    void display()
    {
        //			cout<<endl<<"Your HashMap -- ";
        cout << endl;
        for (int i = 0; i < capacity; i++)
        {
            cout << i << " -- ";
            if (arr[i] != NULL)
            {
                cout << arr[i]->key << " -> " << arr[i]->value;
            }
            cout << endl;
        }
    }

    // function to see collision

    void seeCollision()
    {
        cout << this->collisions;
    }
};

// Print function

void print(string message)
{
    cout << endl
         << message;
}

int main()
{
    // For Calculating Time

    auto start = high_resolution_clock::now();

    // -- end For Calculating Time
    int capacity;
    print("Enter the capacity for which the key will be moded to and array will crested on this - ");
    cin >> capacity;
    HashMap map(capacity);

    // while for menu --
    bool menu = true;
    while (menu)
    {
        print("Collisions - ");
        map.seeCollision();
        print("Your Hash Map -- ");
        map.display();

        cout << endl;

        print("Enter 1 for Add Key Value operation");
        print("Enter 2 for Delete with key operation");
        print("Enter 3 for Size of Map operation");
        print("Enter 4 for Get value From Map operation");

        print("Enter 5 to exit");
        cout << endl;
        int input;
        cin >> input;
        int data;

        switch (input)
        {
        case 1:
            map.insertForMenu();
            break;
        case 2:
            map.deleteNodeVoid();
            break;
        case 3:
            map.getSize();
            break;
        case 4:
            map.getValue();
            break;

        default:
            menu = false;
        }
    }

    // For calculating Time

    auto stop = high_resolution_clock::now();

    auto duration = duration_cast<microseconds>(stop - start);

    cout << "Time taken by function: "
         << duration.count() << " microseconds" << endl;

    // end For Calculating Time

    return 0;
}

File: Hashing Techniques/Double Hashing/HashmapWithDoubleHashing.java
import java.util.Scanner;

public class HashmapWithDoubleHashing {

    int capacity;
    Entry dummyNode;
    Entry[] table;
    int prime;
    int collision;
    long timeTakentoInsert;
    int size;

    // The Entry class represents a key-value pair in the hash map
    static class Entry{
        int key;
        int value;
        Entry(int key,int value){
            this.key= key;
            this.value= value;
        }
    }

    // Constructor for the hash map
    HashmapWithDoubleHashing(int cap) {
        this.capacity= cap;
        table= new Entry[cap];
        dummyNode= new Entry(-1, -1);
        this.prime= getPrime(cap);
        this.timeTakentoInsert=0;
        this.collision = 0;
        this.size = 0;
    }

    // A helper function to check if a number is prime
    boolean isPrime(int n) {
        if(n==1 || n==0)
            return false;
        for(int i=2;i<n;i++){
            if(n%i==0)
                return false;
        }
        return true;
    }

    // A helper function to get the largest prime number less than a given number
    int getPrime(int cap) {
        cap-=1;
        while(cap>0){
            if (isPrime(cap)){
                return cap;
            }
            cap-=1;
        }
        return -1;
    }

    // The first hash function
    int hash1(int key) {
        return key%capacity;
    }

    // The second hash function
    int hash2(int key) {
        return (prime-(key% prime));
    }

    // Insert a key-value pair into the hash map
    void insert(int key,int value) {

        if (size >= capacity) {
            System.out.println("Capacity is full");
            return;
        }

        // Record the start time for inserting the entry
        long s = System.currentTimeMillis();

        Entry newEntry= new Entry(key,value);

        // Compute the hash values
        int h1= hash1(key);
        int h2= hash2(key);

        // Use double hashing to find an empty slot in the hash table
        while(table[h1]!= null){

            // If the key already exists, update the value
            if(table[h1].key==key){
                table[h1].value= value;
                return;
            }
            
            h1+=h2;
            h1%= capacity;
            collision+=1;

        }

        table[h1]= newEntry;

        // Record the end time for inserting the entry
        long e = System.currentTimeMillis();

        // Update the collision count, size, and time taken to insert
        size += 1;
        timeTakentoInsert += (e - s);
    }

    // Remove a key-value pair from the hash map
    int remove(int key) {
        int hash= hash1(key);
        int i=0;
        while(i<capacity) {
            if(table[hash]!= null && table[hash].key==key){
                Entry temp= table[hash];
                table[hash]= dummyNode;
                size-=1;
                return temp.value;
            }
            hash+=1;
            i+=1;
            hash= hash%capacity;
        }
        return -1;
    }


    // Method to retrieve the value of key passed
    int getValue(int key) {
        int hash = hash1(key);
        int h2 = hash2(key);
        int count = 0;
        while (table[hash] != null) {
            if (table[hash].key == key) {
                return table[hash].value;
            }
            hash += h2;
            hash %= capacity;
            count++;
            if (count == capacity) {
                break;
            }
        }
        return -1;
    }
    
    // Method to search if a key is present or not
    boolean search(int key) {
        int h1 = hash1(key);
        int h2 = hash2(key);
        int count = 0;
        while (table[h1] != null) {
            if (table[h1].key == key) {
                return true;
            }
            h1 += h2;
            h1 %= capacity;
            count++;
            if (count == capacity) {
                break;
            }
        }
        return false;
    }

    // Method to display the hash table
    void display() {
        for(int i=0; i<capacity;i++) {
            if(table[i]!= null) {
                System.out.println(i + ":" +   "(" + table[i].key +"," + table[i].value + ")" );
            } else {
                System.out.println(i+ ":");
            }
        }
    }

    // Returns the total no of collision
   int getCollision()
   {
       return collision;
   }


   

   // Returns the time taken to insert
   long getTimetaken(){

       return timeTakentoInsert;

   }



    public static void main(String[] args) {
        try (// Sample usage
        Scanner scanner = new Scanner(System.in)) {
            System.out.println("Enter capacity of HashMap:");
            int capacity = scanner.nextInt();
            HashmapWithDoubleHashing map = new HashmapWithDoubleHashing(capacity);

            while (true) {
                System.out.println("\nChoose an option:");
                System.out.println("1. Insert a key-value pair");
                System.out.println("2. Remove a key-value pair");
                System.out.println("3. Get the value associated with a key");
                System.out.println("4. Search for a key");
                System.out.println("5. No of collisions");
                System.out.println("6. Time taken");
                System.out.println("7. Display the contents of the HashMap");
                System.out.println("8. Exit");

                int option = scanner.nextInt();
                switch (option) {
                    case 1:
                        System.out.println("Enter key:");
                        int key = scanner.nextInt();
                        System.out.println("Enter value:");
                        int value = scanner.nextInt();
                        map.insert(key, value);
                        System.out.println("Key-value pair inserted successfully.");
                        break;
                    case 2:
                        System.out.println("Enter key:");
                        key = scanner.nextInt();
                        value = map.remove(key);
                        if (value == -1) {
                            System.out.println("Key not found.");
                        } else {
                            System.out.println("Value removed: " + value);
                        }
                        break;
                    case 3:
                        System.out.println("Enter key:");
                        key = scanner.nextInt();
                        value = map.getValue(key);
                        if (value == -1) {
                            System.out.println("Key not found.");
                        } else {
                            System.out.println("Value: " + value);
                        }
                        break;
                    case 4:
                        System.out.println("Enter key:");
                        key = scanner.nextInt();
                        boolean found = map.search(key);
                        if (found) {
                            System.out.println("Key found.");
                        } else {
                            System.out.println("Key not found.");
                        }
                        break;

                    case 5:
                    System.out.println(map.getCollision());
                    break;

    
                    case 6:
                    System.out.println(map.getTimetaken());
                    break;

                    case 7:
                        map.display();
                        break;

                    case 8:
                        System.exit(0);
                    default:
                        System.out.println("Invalid option.");
                }
            }
        }
    }
}


